### PR TITLE
MAINT: support python 3.10

### PR DESCRIPTION
**Pull request by tacaswell**
In https://github.com/python/cpython/pull/20290 CPython changed
`Py_TYPE` from a macro to an inline function.  This requires a code
change to us `Py_SET_TYPE` instead when using `Py_TYPE()` as a lvalue
in c code.

~This is a draft (and I expect it to fail all of the tests) because `Py_SET_TYPE` was only introduced in py39.  Can I please have some guidance on how to do version gating (or have someone who knows take over this PR ;) )~

### CODE REVIEW

**Comment by charris** [comment]
Ugh. Is it certain that this change will persist to release?

**Comment by eric-wieser** [comment]
Something like this in `npy3k.h` (or however it's spelt) ]would do it:

```
#if HEX_VERSION < 0x031000
    #define Py_SET_TYPE(obj, typ) (Py_TYPE(obj) = typ)
#endif
```

**Comment by tacaswell** [comment]
This is part of a series of intentional API breaks to try and condense the python c-api so I am pretty confident that this will persist.

**Comment by seberg** [comment]
It seems like we should also replace all `ob_type` then... But, I guess no need to try and get ahead of Python here...

**Comment by charris** [comment]
> This is part of a series of intentional API breaks

We might want to make a compatibility include file to use until Python3.9 gets dropped.  Probably need to use new names to avoid redefinitions.

File: numpy/core/include/numpy/npy_3kcompat.h [hunk 1]
```diff
@@ -546,4 +546,9 @@ NpyCapsule_Check(PyObject *ptr)
 }
 #endif
 
+// account for changing PyType
+#if PY_VERSION_HEX < 0x03090000
```
Comment (eric-wieser) [inline numpy/core/include/numpy/npy_3kcompat.h:550]:
```suggestion
/* Introduced in https://github.com/python/cpython/commit/d2ec81a8c99796b51fb8c49b77a7fe369863226f */
#if PY_VERSION_HEX < 0x030900a4
```

**Comment by charris** [comment]
We are running tests against 3.9-dev, so that should validate that this works.

File: numpy/core/include/numpy/npy_3kcompat.h [hunk 2]
```diff
@@ -546,4 +546,10 @@ NpyCapsule_Check(PyObject *ptr)
 }
 #endif
 
+/* Introduced in https://github.com/python/cpython/commit/d2ec81a8c99796b51fb8c49b77a7fe369863226f */
```
Comment (charris) [inline numpy/core/include/numpy/npy_3kcompat.h:549]: This might look better up top with the similar compatibility hack.

Comment (tacaswell) [reply numpy/core/include/numpy/npy_3kcompat.h:549]: I should also add a reference for the other one.

**Comment by charris** [comment]
The testing error is on the azure end and may be ignored.

**Review by charris — APPROVED** [review]

**Comment by charris** [comment]
Thanks Thomas.

**Pull request merged**